lib/transaction_files.c

```diff
@@ -777,9 +786,9 @@ xbps_transaction_files(struct xbps_handle *xhp, xbps_object_iterator_t iter)
 	while ((obj = xbps_object_iterator_next(iter)) != NULL) {
 		bool update = false;
 		/*
-		 * `idx` is used as package install index, to chose which
-		 * choose the first package which owns or used to own the
-		 * file deletes it.
+		 * `idx` is used as package install index, to choose which
+		 * choose the first or last package which owns or used to
+		 * own the file or directory deletes it.
```

Comment:
I'm not sure I understand this comment. The language is still a bit unclear.

Reply:
I understand idx is related to file removal, but the logic for the removal isn't very explicit.

Reply:
I think the comment as it is is a bit out of scope for `idx`, I can add a toplevel comment to explain how everything works instead.

* `idx` is the packages index in the transaction, the order in which packages are removed and installed.
* for new files, symlinks and directories the first package has to remove the old file to avoid problems with file type changes, i.e. symlink to directory and to avoid that a package later in the transaction which previously owned the file removes the file which is now owned by the previously unpacked package.
* for obsolete files only the order of removal of directories is important, the directory has to be removed after the last file inside the directory has been removed, to avoid issues if multiple package updates/removals own files in that directory.

Reply:
I think copy-pasting your comment there would work well, thanks!

Perhaps mention how it works if the user also added a file in the directory themselves (although I don't think there are many cases where this is *supposed* to happen).

Reply:
> Perhaps mention how it works if the user also added a file in the directory themselves (although I don't think there are many cases where this is supposed to happen).

* If the directory is just obsolete it will just do nothing about it and leave user file and directory behind.
* If the directory is supposed to be replaced with a regular file or a symlink, it needs to be able to delete directory, in this case it will count how many files are being removed and how many files are there. If there are more files than it can remove, xbps will abort the transaction before unpacking any packages. The whole file check happens after downloading and verifying binary packages (it needs the files list from the packages metadata) but before any other change is done to the system or the package database.

Reply:
Got it, I think. Thanks a lot!